src/java.base/share/classes/sun/security/jca/Providers.java

```diff
@@ -90,14 +90,18 @@ private Providers() {
     // See there for details.
 
     // Hardcoded names of providers to use for JAR verification.
+    //
+    // The fully qualified names of the providers are, also, supplied to
+    // allow loading them when RestrictedSecurity is enabled.
+    //
     // MUST NOT be on the bootclasspath and not in signed JAR files.
     private static final String[] jarVerificationProviders = {
-        "SUN",
-        "SunRsaSign",
+        "SUN", "sun.security.provider.Sun",
+        "SunRsaSign", "sun.security.rsa.SunRsaSign",
         // Note: when SunEC is in a signed JAR file, it's not signed
         // by EC algorithms. So it's still safe to be listed here.
-        "SunEC",
-        "SunJCE",
+        "SunEC", "sun.security.ec.SunEC",
+        "SunJCE", "com.sun.crypto.provider.SunJCE",
     };
```

Comment:
To avoid potentially loading providers twice, perhaps we should check `RestrictedSecurity.isEnabled()` to decide whether to use a list of fully-qualified names instead of the normal list?

Reply:
Changed to have two separate arrays and a check to decide which one to use.